internal/aggregated-co/writer.go

```diff
@@ -0,0 +1,87 @@
+package aggregated_co
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+)
+
+// From https://github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorstatus/writer.go
+
+// NewWriter returns a new instance of Writer.
+func NewWriter(discovery discovery.DiscoveryInterface, client configv1client.ConfigV1Interface) *Writer {
+	return &Writer{
+		discovery: discovery,
+		client:    client,
+	}
+}
+
+// Writer encapsulates logic for cluster operator object API. It is used to
+// update ClusterOperator resource.
+type Writer struct {
+	discovery discovery.DiscoveryInterface
+	client    configv1client.ConfigV1Interface
+}
+
+// EnsureExists ensures that the cluster operator resource exists with a default
+// status that reflects expecting status.
+func (w *Writer) EnsureExists(name string) (existing *configv1.ClusterOperator, err error) {
+	existing, err = w.client.ClusterOperators().Get(context.TODO(), name, metav1.GetOptions{})
+	if err == nil {
+		return
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return
+	}
+
+	co := &configv1.ClusterOperator{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+	existing, err = w.client.ClusterOperators().Create(context.TODO(), co, metav1.CreateOptions{})
+	return
+}
+
+// UpdateStatus updates the clusteroperator object with the new status specified.
+func (w *Writer) UpdateStatus(existing *configv1.ClusterOperator, newStatus *configv1.ClusterOperatorStatus) error {
+	if newStatus == nil || existing == nil {
+		return errors.New("input specified is <nil>")
```

Comment:
Should the function signature use concrete values rather than pointers so that we can avoid this `nil` check. 

As an aside, this error seems more like a programmer error than a user error, so perhaps a `panic` is more appropriate than returning it and re-reconciling forever?

Reply:
This came from the OLM project, I can update it